File: snippets/go/develop/durablepromise.go
package develop

import (
	restate "github.com/restatedev/sdk-go"
	"log/slog"
)

// <start_review>
type ReviewWorkflow struct{}

func (ReviewWorkflow) Run(ctx restate.WorkflowContext, documentId string) (string, error) {
	// Send document for review
	if _, err := restate.Run(ctx, func(ctx restate.RunContext) (restate.Void, error) {
		return restate.Void{}, askReview(documentId)
	}); err != nil {
		return "", err
	}

	// Wait for external review submission
	// <start_promise>
	review, err := restate.Promise[string](ctx, "review").Result()
	// <end_promise>
	if err != nil {
		return "", err
	}

	// Process the review result
	return processReview(documentId, review)
}

func (ReviewWorkflow) SubmitReview(ctx restate.WorkflowSharedContext, review string) error {
	// Signal the waiting run handler
	return restate.Promise[string](ctx, "review").Resolve(review)
}

// <end_review>

func (ReviewWorkflow) SubmitReviewClean(ctx restate.WorkflowSharedContext, review string) error {
	// To have a clean example without return in front of it
	// <start_resolve_promise>
	err := restate.Promise[string](ctx, "review").Resolve(review)
	if err != nil {
		return err
	}
	// <end_resolve_promise>
	return nil
}

func processReview(id string, review string) (string, error) {
	// Process the review result and return a message
	return "Processed review for document " + id + ": " + review, nil
}

func askReview(id string) error {
	slog.Info("Asking for review", "documentId", id)
	// Simulate asking for a review, e.g., sending an email or notification
	return nil
}

File: snippets/python/src/develop/durable_promise.py
import restate

# <start_review>
review_workflow = restate.Workflow("ReviewWorkflow")


@review_workflow.main()
async def run(ctx: restate.WorkflowContext, document_id: str):
    # Send document for review
    await ctx.run_typed("ask review", ask_review, document_id=document_id)

    # Wait for external review submission
    # <start_promise>
    review = await ctx.promise("review", type_hint=str).value()
    # <end_promise>

    # Process the review result
    return process_review(document_id, review)


@review_workflow.handler()
async def submit_review(ctx: restate.WorkflowSharedContext, review: str):
    # Signal the waiting run handler
    # <start_resolve_promise>
    await ctx.promise("review", type_hint=str).resolve(review)
    # <end_resolve_promise>


app = restate.app([review_workflow])
# <end_review>


def ask_review(document_id: str) -> str:
    # Simulate asking for a review
    return f"Review requested for document {document_id}"


def process_review(document_id: str, review: str) -> str:
    # Simulate processing the review
    return f"Processed review '{review}' for document {document_id}"
